demos/keyexch/x25519.c

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+
+/*
+ * This is a demonstration of key exchange using X25519.
+ *
+ * The variables beginning `peer1_` / `peer2_` are data which would normally be
+ * accessible to that peer.
+ *
+ * Ordinarily you would use random keys, which are demonstrated
+ * below when use_kat=0. A known answer test is demonstrated
+ * when use_kat=1.
+ */
+
+/* A property query used for selecting the X25519 implementation. */
+static const char *propq = NULL;
+
+static const unsigned char peer1_privk_data[32] = {
+    0x80, 0x5b, 0x30, 0x20, 0x25, 0x4a, 0x70, 0x2c,
+    0xad, 0xa9, 0x8d, 0x7d, 0x47, 0xf8, 0x1b, 0x20,
+    0x89, 0xd2, 0xf9, 0x14, 0xac, 0x92, 0x27, 0xf2,
+    0x10, 0x7e, 0xdb, 0x21, 0xbd, 0x73, 0x73, 0x5d
+};
+
+static const unsigned char peer2_privk_data[32] = {
+    0xf8, 0x84, 0x19, 0x69, 0x79, 0x13, 0x0d, 0xbd,
+    0xb1, 0x76, 0xd7, 0x0e, 0x7e, 0x0f, 0xb6, 0xf4,
+    0x8c, 0x4a, 0x8c, 0x5f, 0xd8, 0x15, 0x09, 0x0a,
+    0x71, 0x78, 0x74, 0x92, 0x0f, 0x85, 0xc8, 0x43
+};
+
+static const unsigned char expected_result[32] = {
+  0x19, 0x71, 0x26, 0x12, 0x74, 0xb5, 0xb1, 0xce,
+  0x77, 0xd0, 0x79, 0x24, 0xb6, 0x0a, 0x5c, 0x72,
+  0x0c, 0xa6, 0x56, 0xc0, 0x11, 0xeb, 0x43, 0x11,
+  0x94, 0x3b, 0x01, 0x45, 0xca, 0x19, 0xfe, 0x09
+};
+
+typedef struct peer_data_st {
+    const char *name;               /* name of peer */
+    EVP_PKEY *privk;                /* privk generated for peer */
+    unsigned char pubk_data[32];    /* generated pubk to send to other peer */
+
+    unsigned char *secret;          /* allocated shared secret buffer */
+    size_t secret_len;
+} PEER_DATA;
+
+/*
+ * Prepare for X25519 key exchange. The public key to be sent to the remote peer
+ * is put in pubk_data, which should be a 32-byte buffer. Returns 1 on success.
+ */
+static int keyexch_x25519_before(
+    OSSL_LIB_CTX *libctx,
+    const unsigned char *kat_privk_data,
+    PEER_DATA *local_peer)
+{
+    int rv = 0;
+    size_t pubk_data_len = 0;
+
+    /* Generate or load X25519 key for the peer */
+    if (kat_privk_data)
+        local_peer->privk = EVP_PKEY_new_raw_private_key_ex(libctx, "X25519", propq,
+            kat_privk_data, sizeof(peer1_privk_data));
```

Comment:
Sorry didn't spot this on the first run through...I'd split this like so:

```` C
        local_peer->privk =
            EVP_PKEY_new_raw_private_key_ex(libctx, "X25519", propq,
                                            kat_privk_data,
                                            sizeof(peer1_privk_data));
````

We generally try to line things up with the paren. The coding style is a bit vague on this though:

"Statements longer than 80 columns must be broken into sensible chunks, unless exceeding 80 columns significantly increases readability and does not hide information. Descendants are always substantially shorter than the parent and are placed substantially to the right."

There are several other places like this. I'd review all line breaks with this in mind.